beacon_node/beacon_chain/src/execution_proof_verification.rs

```diff
@@ -0,0 +1,625 @@
+/// network.
+pub struct GossipVerifiedExecutionProof<T: BeaconChainTypes, O: ObservationStrategy = Observe> {
+    block_root: Hash256,
+    execution_proof: Arc<ExecutionProof>,
+    _phantom: PhantomData<(T, O)>,
+}
+
+impl<T: BeaconChainTypes, O: ObservationStrategy> std::fmt::Debug
+    for GossipVerifiedExecutionProof<T, O>
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("GossipVerifiedExecutionProof")
+            .field("block_root", &self.block_root)
+            .field("execution_proof", &self.execution_proof)
+            .finish()
+    }
+}
+
+impl<T: BeaconChainTypes, O: ObservationStrategy> Clone for GossipVerifiedExecutionProof<T, O> {
+    fn clone(&self) -> Self {
+        Self {
+            block_root: self.block_root,
+            execution_proof: self.execution_proof.clone(),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T: BeaconChainTypes, O: ObservationStrategy> GossipVerifiedExecutionProof<T, O> {
+    pub fn new(
+        execution_proof: Arc<ExecutionProof>,
+        chain: &BeaconChain<T>,
+    ) -> Result<Self, GossipExecutionProofError> {
+        validate_execution_proof_for_gossip::<T, O>(execution_proof, chain)
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.execution_proof.slot
+    }
+
+    pub fn block_root(&self) -> Hash256 {
+        self.block_root
+    }
+
+    pub fn execution_proof(&self) -> &Arc<ExecutionProof> {
+        &self.execution_proof
+    }
+
+    pub fn subnet_id(&self) -> ExecutionProofId {
+        self.execution_proof.proof_id
+    }
+
+    /// Get the block root for this proof.
+    pub fn into_inner(self) -> Arc<ExecutionProof> {
+        self.execution_proof
+    }
+}
+
+/// Validate an execution proof for gossip
+pub fn validate_execution_proof_for_gossip<T: BeaconChainTypes, O: ObservationStrategy>(
+    execution_proof: Arc<ExecutionProof>,
+    chain: &BeaconChain<T>,
+) -> Result<GossipVerifiedExecutionProof<T, O>, GossipExecutionProofError> {
+    let block_root = execution_proof.block_root;
+    let proof_slot = execution_proof.slot;
+
+    // 1. Verify proof is not from the future
+    verify_proof_not_from_future_slot(chain, proof_slot)?;
+
+    // 2. Verify proof slot is greater than finalized slot
+    verify_slot_greater_than_latest_finalized_slot(chain, proof_slot)?;
+
+    // 3. Check if proof is already known via gossip
+    verify_is_unknown_execution_proof(chain, &execution_proof)?;
+
+    // 4. Check if the proof is already in the DA checker cache
+    // If it exists in the cache, we know it has already passed validation.
+    if chain
+        .data_availability_checker
+        .is_execution_proof_cached(&block_root, &execution_proof)
+    {
+        if O::observe() {
+            observe_gossip_execution_proof(&execution_proof, chain)?;
+        }
+        return Err(GossipExecutionProofError::PriorKnownUnpublished);
```

Comment:
Gossip has a de-duplication mechanism based on the message content. Is there a single possible `ExecutionProof` per subnet and slot that is valid? If yes, the observe cache feels unnecessary.